### ISSUE TITLE
vacuum hanging when passing libopenapi.Document into `ApplyRulesToRuleSet`

### ISSUE THREAD

**Issue by TristanSpeakEasy** [daveshanley/vacuum#348]
This code reproduces the issue 

```
package main

import (
    "fmt"
    "os"

    "github.com/pb33f/libopenapi"
    "github.com/pb33f/libopenapi/datamodel"

    "github.com/daveshanley/vacuum/model"
    "github.com/daveshanley/vacuum/motor"
    "github.com/daveshanley/vacuum/rulesets"
)

func read() []byte {
    data, err := os.ReadFile("test.yaml")
    if err != nil {
        panic(err)
    }

    return data
}

func main() {
    config := datamodel.DocumentConfiguration{
        IgnorePolymorphicCircularReferences: true,
        IgnoreArrayCircularReferences:       true,
    }

    doc, err := libopenapi.NewDocumentWithConfiguration(read(), &config)
    if err != nil {
        panic(err)
    }

    ex := &motor.RuleSetExecution{
        RuleSet: &rulesets.RuleSet{
            Rules: make(map[string]*model.Rule),
        },
        Document: doc,
        CustomFunctions: make(map[string]model.RuleFunction),
        Base:            "./",
    }

    motor.ApplyRulesToRuleSet(ex)

    fmt.Println("done")
}
```

OpenAPI documents required to create this bug already provided to @daveshanley privately due to sensitive nature.

**Comment by daveshanley** [comment]
This bug has exposed some major deficiencies in `libopenapi`'s indexing design. To fix this correctly, we need a new feature. The `rolodex` is coming to town.

**Comment by daveshanley** [comment]
Fixed with `v0.4` of vacuum. The rolodex is in play and working well.
